src/server/pps/server/worker_pool.go

```diff
@@ -195,36 +205,80 @@ func (w *workerPool) addWorker(addr string) error {
 
 	wr := &worker{
 		ctx:          childCtx,
+		cancel:       cancelFn,
 		addr:         addr,
 		workerClient: workerpkg.NewWorkerClient(conn),
 		pachClient:   pachClient,
+		retries:      w.retries,
 	}
+	w.workersMapMu.Lock()
+	w.workersMap[addr] = wr
+	w.workersMapMu.Unlock()
 	protolion.Infof("launching new worker at %v", addr)
 	go wr.run(w.dataCh)
 	return nil
 }
 
 func (w *workerPool) delWorker(addr string) error {
-	cancel, ok := w.workersMap[addr]
+	w.workersMapMu.RLock()
+	defer w.workersMapMu.RUnlock()
+	worker, ok := w.workersMap[addr]
 	if !ok {
 		return fmt.Errorf("deleting worker %s which is not in worker pool", addr)
 	}
-	cancel()
+	worker.cancel()
 	return nil
 }
 
 func (w *workerPool) DataCh() chan *datumAndResp {
 	return w.dataCh
 }
 
+func status(ctx context.Context, id string, etcdClient *etcd.Client, etcdPrefix string) ([]*pps.WorkerStatus, error) {
+	workerClients, err := workerClients(ctx, id, etcdClient, etcdPrefix)
+	if err != nil {
+		return nil, err
+	}
+	var result []*pps.WorkerStatus
+	for _, workerClient := range workerClients {
+		status, err := workerClient.Status(ctx, &types.Empty{})
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, status)
+	}
+	return result, nil
+}
+
+func cancel(ctx context.Context, id string, etcdClient *etcd.Client,
+	etcdPrefix string, jobID string, dataFilter []string) error {
+	workerClients, err := workerClients(ctx, id, etcdClient, etcdPrefix)
+	if err != nil {
+		return err
+	}
+	for _, workerClient := range workerClients {
+		status, err := workerClient.Status(ctx, &types.Empty{})
+		if err != nil {
+			return err
+		}
+		if jobID == status.JobID && workerpkg.MatchDatum(dataFilter, status.Data) {
```

Comment:
Why are we checking `jobID` and `dataFilter` here?  Looking at the implementation of `Cancel`, it already checks `dataFilter`.  Also, checking `jobID` on the server side might not be enough, since there can be a race where the worker moves on to process datum for another job after you called `Status`.  Maybe `Cancel` should take a `jobID` and the worker (instead of the server) can check that the job matches?